File: clients/python/examples/with_payload.py
"""Beat loop that packs queue depth and last error code into the 32-bit payload.

Layout: high 16 bits = queue_depth, low 16 bits = last_error_code.
The observer treats the payload opaquely; decoding is the agent's concern.

Run alongside varta-watch::

    varta-watch --socket /tmp/varta.sock --threshold-ms 2000 &
    python with_payload.py /tmp/varta.sock
"""

from __future__ import annotations

import sys
import time

from varta import Status, Varta


def main(argv: list[str]) -> int:
    path = argv[1] if len(argv) > 1 else "/tmp/varta.sock"
    queue_depth = 0
    last_error = 0
    with Varta.connect(path) as agent:
        while True:
            payload = ((queue_depth & 0xFFFF) << 16) | (last_error & 0xFFFF)
            outcome = agent.beat(Status.OK, payload)
            if outcome.is_dropped:
                print(f"varta: beat dropped ({outcome.reason.value})", file=sys.stderr)
            time.sleep(0.5)


if __name__ == "__main__":
    sys.exit(main(sys.argv))

File: clients/node/examples/with_payload.ts
// Beat loop that packs queue depth and last error code into the 32-bit
// payload field. Mirror of `clients/python/examples/with_payload.py`.
//
// Wire layout for the 32-bit payload here:
//   bits  0..15  = queue depth (u16)
//   bits 16..23  = last error code (u8)
//   bits 24..31  = reserved (0)
//
// The observer dashboard at `observability/dashboards/varta-health.json`
// surfaces the payload column verbatim, so any encoding is fine as long
// as your team agrees on it.
//
// Usage:
//   node --import tsx examples/with_payload.ts [host=127.0.0.1] [port=5876]

import { Varta, Status } from "../src/index.js";

const host = process.argv[2] ?? "127.0.0.1";
const port = parseInt(process.argv[3] ?? "5876", 10);

const agent = Varta.connectUdp(host, port);

function packPayload(queueDepth: number, lastErrorCode: number): number {
  return ((queueDepth & 0xffff) | ((lastErrorCode & 0xff) << 16)) >>> 0;
}

let depth = 0;
let lastError = 0;

const timer = setInterval(() => {
  // Pretend metrics. In production these come from your queue/observability layer.
  depth = (depth + 1) & 0xffff;
  lastError = depth % 4 === 0 ? 1 : 0;

  const status = lastError !== 0 ? Status.Degraded : Status.Ok;
  agent.beat(status, packPayload(depth, lastError));
}, 500);

const shutdown = (): void => {
  clearInterval(timer);
  agent.close();
  process.exit(0);
};
process.on("SIGINT", shutdown);
process.on("SIGTERM", shutdown);
